### ISSUE TITLE
DOC: Provide a post-processing function to replace ligatures

### ISSUE THREAD

**Issue by MartinThoma** [py-pdf/pypdf#2046]
`pypdf==3.14.0` extracts ligatures as their unicode symbols, e.g. `ﬀ` is one unicode symbol. Many users might expect `ff` (two unicode symbols) instead.

`pypdf` could provide a post-processing option for all ligatures https://en.wikipedia.org/wiki/Ligature_(writing)#Ligatures_in_Unicode_(Latin_alphabets)

## Is post-processing desirable?

I think most users would expect this post-processing, because I think people mostly use it for NLP-like tasks:

* Understand what the PDF is about
* Summarize a PDF
* Extract specific information from the PDF
* Text-to-speech (TTS): Let software read the PDF out loud. This might also be relevant for people with disabilities, although they likely have some specialized software

Use-cases where it's not desirable:

* The document itself could be about ligatures, e.g. about typesetting
* The document could be about Unicode
* The document could be about text processing / search

I cannot imagine any other case where we would NOT want to replace ligatures by their single characters.

## How we could implement it

We could introduce a new parameter `post_processing` to `extract_text` that expects a list of functions. We would add `pypdf/post_processing.py`:

```
def ligature_removal(text_in: str) -> str:
    # this needs to be expanded
    return text_in.replace("ﬁ", "ff")

# other functions could be: Page number removal, de-hyphenation, thousands separator detection
```

example:

```python
from pypdf import PdfReader

reader = PdfReader("input.pdf")
page = reader.pages[0].

# applies post processing by default as I think most people want it:
page.extract_text()  

# but give the user the option to disable it / add their own:
page.extract_text(post_processing=[])
```

**Comment by MartinThoma** [comment]
Instead of simple functions we could use objects. That gives us potentially more flexibility.

I'm thinking a bit of pipelines as used by sklearn: https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html

We could allow those PostProcessing objects to have two functions:

* `fit`: Receive the whole extracted (raw) text
* `extract`: Receive the input of the step before and supply it to the next

Most functions will likely not do anything in `fit`, but a function that tries to get rid of thousands separators might want to detect the documents language first.

**Comment by MartinThoma** [comment]
I've added the https://github.com/py-pdf/pypdf/labels/help%20wanted mainly to get feedback:

1. **Scope of pypdf**: Should something like this go into pypdf in the first place? It feels a lot more like NLP and the PDF-specific parts might not be that high. Likely the Spacy / nltk community already has good post-processing for PDFs. We should not re-invent the wheel. Maybe we can just point to one of those projects and add a documentation page given an example how to use it?
2. **Design**: If it should go in: Which structure would you prefer?

**Comment by MartinThoma** [comment]
To emphasize: **PLEASE DONT CREATE PRs for this issue!** This still needs discussion. I'll let you know once I came to a conclusion if it should go in / what it should look like :-)

**Comment by pubpub-zz** [comment]
the more I think about, the more I come to the idea that we will need a dictionnary to identify when to remove hyphen or not.

**Comment by MartinThoma** [comment]
You mean a dictionary of words, right? So that we can just distinguish between word-breaks and the other usages of dashes: https://en.m.wikipedia.org/wiki/Dash#En_dash

That would cover most cases, but not all. For example, new words get created all the time. Googling vs Go ogling. Also there are many languages. Way more than there are countries.

**Comment by MartinThoma** [comment]
The more I think about it, the more I come to the conclusion that this is out of scope for pypdf 😁 it is nlp.

However, I think we should document how it can be done.

**Comment by stefan6419846** [comment]
Given my use case which includes building search indices, adding an option to enable/disable ligature splitting seems like a valid idea. Otherwise, this might require post-processing on my side, as I do not expect (regular) users to enter actual ligatures into a search field.

**Comment by stefan6419846** [comment]
> The action to fix an hyphen should be `extracted_text.replace("-\n","")` then a toggle may not be so useful.

This ticket is mostly about ligatures, not about hyphenation, and this is what my comment refers to as well.

**Comment by pubpub-zz** [comment]
Oups... But still valid with other parameters😚
